File: src/main/java/com/zunza/buythedip/auth/oauth2/CustomOAuth2UserService.java
package com.zunza.buythedip.auth.oauth2;

import java.util.UUID;

import org.springframework.context.ApplicationEventPublisher;
import org.springframework.security.oauth2.client.userinfo.DefaultOAuth2UserService;
import org.springframework.security.oauth2.client.userinfo.OAuth2UserRequest;
import org.springframework.security.oauth2.core.OAuth2AuthenticationException;
import org.springframework.security.oauth2.core.user.OAuth2User;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.zunza.buythedip.auth.oauth2.exception.DuplicateEmailWithDifferentProviderException;
import com.zunza.buythedip.auth.oauth2.exception.SocialEmailAlreadyRegisteredException;
import com.zunza.buythedip.auth.oauth2.dto.GoogleOAuth2Response;
import com.zunza.buythedip.auth.oauth2.dto.KakaoOAuth2Response;
import com.zunza.buythedip.auth.oauth2.dto.NaverOAuth2Response;
import com.zunza.buythedip.auth.oauth2.dto.OAuth2Response;
import com.zunza.buythedip.infrastructure.event.event.UserRegisteredEvent;
import com.zunza.buythedip.user.constant.UserType;
import com.zunza.buythedip.user.entity.User;
import com.zunza.buythedip.user.repository.UserRepository;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class CustomOAuth2UserService extends DefaultOAuth2UserService {
	private final UserRepository userRepository;
	private final ApplicationEventPublisher eventPublisher;

	@Override
	@Transactional
	public OAuth2User loadUser(OAuth2UserRequest userRequest) throws OAuth2AuthenticationException {
		OAuth2User oAuth2User = super.loadUser(userRequest);
		String provider = userRequest.getClientRegistration().getRegistrationId();
		OAuth2Response oAuth2Response = getOAuth2Response(oAuth2User, provider);

		return userRepository.findByEmail(oAuth2Response.getEmail())
			.map(user -> {
				validateExistingUserForOAuth2Login(user, oAuth2Response);
				return new CustomOAuth2User(user);
			})
			.orElseGet(() -> {
				User user = User.createSocialUser(oAuth2Response, createRandomNickname());
				User savedUser = userRepository.save(user);

				eventPublisher.publishEvent(UserRegisteredEvent.createFrom(savedUser));
				return new CustomOAuth2User(user);
			});
	}

	private OAuth2Response getOAuth2Response(
		OAuth2User oAuth2User,
		String provider
	) {
		return switch (provider) {
			case "google" -> new GoogleOAuth2Response(oAuth2User.getAttributes());
			case "naver" -> new NaverOAuth2Response(oAuth2User.getAttributes());
			case "kakao" -> new KakaoOAuth2Response(oAuth2User.getAttributes());
			default -> throw new IllegalArgumentException("지원하지 않는 Provider 입니다.");
		};
	}

	private void validateExistingUserForOAuth2Login(
		User user,
		OAuth2Response oAuth2Response
	) {
		if (user.getUserType() == UserType.NORMAL) {
			throw new SocialEmailAlreadyRegisteredException();
		}

		if (user.getOAuth2Provider() != oAuth2Response.getProvider()) {
			throw new DuplicateEmailWithDifferentProviderException();
		}
	}

	private String createRandomNickname() {
		return "user" + UUID.randomUUID().toString().substring(0, 8);
	}
}

File: src/main/kotlin/com/zunza/buythedip/auth/oauth2/CustomOAuth2UserService.kt
package com.zunza.buythedip.auth.oauth2

import com.zunza.buythedip.auth.oauth2.dto.GoogleOAuth2Response
import com.zunza.buythedip.auth.oauth2.dto.KakaoOAuth2Response
import com.zunza.buythedip.auth.oauth2.dto.NaverOAuth2Response
import com.zunza.buythedip.auth.oauth2.dto.OAuth2Response
import com.zunza.buythedip.auth.oauth2.exception.DuplicateEmailWithDifferentProviderException
import com.zunza.buythedip.auth.oauth2.exception.SocialEmailAlreadyRegisteredException
import com.zunza.buythedip.user.constant.UserType
import com.zunza.buythedip.user.entity.User
import com.zunza.buythedip.user.repository.UserRepository
import org.springframework.security.oauth2.client.userinfo.DefaultOAuth2UserService
import org.springframework.security.oauth2.client.userinfo.OAuth2UserRequest
import org.springframework.security.oauth2.core.user.OAuth2User
import org.springframework.stereotype.Component
import java.util.UUID

@Component
class CustomOAuth2UserService(
    private val userRepository: UserRepository
) : DefaultOAuth2UserService() {

    override fun loadUser(userRequest: OAuth2UserRequest?): OAuth2User? {
        val oAuth2User = super.loadUser(userRequest)
        val provider = userRequest?.clientRegistration?.registrationId
        val oAuth2Response = getOAuth2Response(oAuth2User, provider)

        return userRepository.findByEmail(oAuth2Response.getEmail())
            ?.let { user ->
                validateExistingUserForOAuth2Login(user, oAuth2Response)
                CustomOAuth2User.createOf(user)
            }
            ?: run {
                val user = User.createSocialUser(oAuth2Response, createRandomNickname())
                val savedUser = userRepository.save(user)
                CustomOAuth2User.createOf(savedUser)
            }
    }

    private fun getOAuth2Response(
        oAuth2User: OAuth2User,
        provider: String?
    ): OAuth2Response {
        return when (provider) {
            "google" -> GoogleOAuth2Response(oAuth2User.attributes)
            "kakao" -> KakaoOAuth2Response(oAuth2User.attributes)
            "naver" -> NaverOAuth2Response(oAuth2User.attributes)
            else -> throw IllegalArgumentException("지원하지 않는 Provider 입니다.")
        }
    }

    private fun validateExistingUserForOAuth2Login(user: User, oAuth2Response: OAuth2Response) {
        when {
            user.type == UserType.NORMAL ->
                throw SocialEmailAlreadyRegisteredException()
            user.provider != oAuth2Response.getProvider() ->
                throw DuplicateEmailWithDifferentProviderException()
        }
    }

    private fun createRandomNickname(): String {
        return "user${UUID.randomUUID().toString().substring(0, 8)}"
    }
}
